### PR TITLE
Made new config for e2e tests having a non alpine image with cgo enabled

### PR DESCRIPTION
**Pull request by tend2infinity**
* [ ] I added CHANGELOG entry for this change.
* [x] Change is not relevant to the end user.

## Changes

- Added a new Dockerfile specifically for performing e2e tests using a non alpine image
- Added a new promu configuration for enabling cgo and race flags
- Specified targets in the makefile

## Verification
- Tested locally
### Part of
Issue #4664

### CODE REVIEW

File: .promue2e.yml
Comment (matej-g) [inline .promue2e.yml:1]: Nit: Maybe we should name this file something like `.promu.e2e.yml`? It looks a bit more readable to me :nerd_face:.

File: Dockerfile.e2e-tests
Comment (matej-g) [inline Dockerfile.e2e-tests:10]: Looks like this was just copy pasted from `Dockerfile.multi-stage` and we don't need it here?

File: Dockerfile.e2e-tests
Comment (matej-g) [inline Dockerfile.e2e-tests:14]: Should this be `build-e2e`? Otherwise we're building binary without the race detector, right?

```suggestion
RUN git update-index --refresh; make build-e2e
```

File: Makefile
Comment (matej-g) [inline Makefile:188]: Do we need to do `build-e2e` if it's part of the Dockerfile build?

**Review by matej-g — COMMENTED** [review]
This looking nicely :clap: I added few comments + just please make sure you signed your commits (see section 'Signing your work' in https://github.com/thanos-io/thanos/blob/main/CONTRIBUTING.md#pull-request-process), so that we can kick start the CI as well!

File: .promue2e.yml
Comment (tend2infinity) [reply .promue2e.yml:1]: Okay, will make this change

File: Dockerfile.e2e-tests
Comment (tend2infinity) [reply Dockerfile.e2e-tests:10]: Yup, actually I wanted to keep the next set of comments here, which were related to the "COPY" command, I guess I've mistakenly removed that and kept this comment, no problems! I'll make the change

File: Dockerfile.e2e-tests
Comment (tend2infinity) [reply Dockerfile.e2e-tests:14]: Tested locally and this works alright!

File: Makefile
Comment (tend2infinity) [reply Makefile:188]: Yes, now this needs to be removed

**Comment by tend2infinity** [comment]
@matej-g made the required changes

**Review by matej-g — COMMENTED** [review]
Looking good, thanks @tend2infinity! Let's ping maintainers to make CI run for this PR.

File: Dockerfile.e2e-tests
Comment (squat) [inline Dockerfile.e2e-tests:1]: We need to update this comment, i think:
```suggestion
# By default we pin to amd64 sha. Use make docker-e2e to automatically adjust for arm64 versions.
```

File: .promu.e2e.yml
Comment (squat) [inline .promu.e2e.yml:21]: Why do we need anything other than Linux for the e2e tests? All of the tests run inside of a container, which means that the will run on Linux, and thus need a Linux binary, right?

File: Makefile
Comment (squat) [inline Makefile:152]: Nit
```suggestion
```

File: Makefile
Comment (squat) [inline Makefile:150]: Why do we need promu at all? No matter what platform i am running the tests on, the way the tests are configured, only a single image will be used, and that image will always have the same architecture as the machine on which `make test-e2e` was invoked.

It seems to me that promu is overkill/overengineered here. Instead, the Dockerfile.e2e-tests should directly embed the `go build` command with all of the necessary flags. cc @matej-g

File: Dockerfile.e2e-tests
Comment (tend2infinity) [reply Dockerfile.e2e-tests:1]: Gotcha! will update this

File: .promu.e2e.yml
Comment (tend2infinity) [reply .promu.e2e.yml:21]: Okay, makes sense , we can remove the other platforms for sure

File: Makefile
Comment (tend2infinity) [reply Makefile:150]: Actually I was following the convention, since the normal build command in the makefile is using promu and just to maintain homogeneity I used the promu utility here as well, maybe we can have a discussion around this in office hours, whatever is the conclusion, I can make the required changes.

Comment (matej-g) [reply Makefile:150]: My thinking was the same, regardless of whether we need multiplatform build or not - since we're using Promu in the normal build flow, I didn't find it to be too much of an overhead, while being more uniform with the 'main' build flow. If the feeling is this is overengineerd, I don't have a problem with switching to simple `go build`.

Comment (tend2infinity) [reply Makefile:150]: I guess it's postponed and maybe we'll have it on feb 3, I'll discuss this for sure and get this unblocked

Comment (matej-g) [reply Makefile:150]: I'm not sure if we'll be able to get a clear guidance, maybe we can try to continue with the solution that seem to be agreeable for each of us, i.e. go with the `go build` suggestion? That would help us to unblock ASAP and we can always iterate later on an alternative.

Either way it looks like unblocking the failing E2E tests would need to be addressed before that.

Comment (tend2infinity) [reply Makefile:150]: @matej-g Okay I'll integrate everything in the docker file itself

Comment (tend2infinity) [reply Makefile]: Hey @matej-g you can check now, all the e2e tests passed locally for me

**Comment by tend2infinity** [comment]
Hey @matej-g can you review it once ;)

**Review by matej-g — DISMISSED** [review]
I think this is looking good :+1: One more cleanup suggestion :broom:

**Comment by tend2infinity** [comment]
Changes done, works alright!

**Comment by matej-g** [comment]
Thanks @tend2infinity and sorry for the delayed answer! Looks like everything is working fine, we are actually already catching some data races. Did you hit these during local runs? I think we should fix them before we merge this (otherwise E2E  tests will be broken). What I found based on test logs on my local machine:
- Data race in label set method on bucket store (I added fix, see PR https://github.com/thanos-io/thanos/pull/5204)
- Data race in receive in remote write handler (somewhere in https://github.com/thanos-io/thanos/blob/main/pkg/receive/handler.go#L538 - needs investigating / fixing)
- Data race in blocks API in set method (https://github.com/thanos-io/thanos/blob/main/pkg/api/blocks/v1.go#L133 - needs investigating / fixing)

**Comment by matej-g** [comment]
Still relevant but not finished, on my to-do list...

**Comment by tend2infinity** [comment]
Hey @matej-g sorry for the delay, I will look into these issues that you've mentioned and will let you know the progress

**Pull request closed without merging**
